### PR TITLE
Mes 3123 search booking with unit tests

### PR DESCRIPTION
**Pull request by amobarak**
# Description and relevant Jira numbers

Search booking microservice complete, with unit tests.
JIRA: https://jira.dvsacloud.uk/browse/MES-3123

## Pull Request checklist

- [x] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop

## Sign off process checklist

- [x] Code has been tested manually
- [ ] Tested by QA
- [ ] PO's approval
- [ ] Reviewers selected in Github
- [ ] PR link added to JIRA

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,10 +25,13 @@
   "dependencies": {
     "@types/jwt-decode": "^2.2.1",
     "aws-sdk": "^2.389.0",
-    "moment": "^2.23.0"
+    "moment": "^2.23.0",
+    "@hapi/joi": "^15.0.3",
+    "@types/hapi__joi": "^15.0.1"
```
Comment (TomBillingtonUK) [inline package.json]: @types - should be a dev dependency

Comment (amobarak) [reply package.json]: Aye sir

File: src/functions/searchBooking/framework/handler.ts [hunk 2]
```diff
@@ -1,6 +1,98 @@
...
+  if (!event.queryStringParameters.appRef) {
+    return createResponse('Query parameter app reference needs to be supplied', HttpStatus.BAD_REQUEST);
+  }
+
+  const applicationReference: string = event.queryStringParameters.appRef;
+
+  const staffNumber = event.pathParameters['staffNumber'];
+
+  const parametersSchema = joi.object().keys({
+    staffNumberValidator: joi.string().alphanum().optional(),
```
Comment (TomBillingtonUK) [inline src/functions/searchBooking/framework/handler.ts]: could this be number

File: src/functions/searchBooking/framework/handler.ts [hunk 3]
```diff
@@ -1,6 +1,98 @@
...
+    bookingSequence:
+      parseInt(applicationReference.substring(applicationReference.length - 3, applicationReference.length - 1), 10),
+  };
+
+  const parameterAppRef: number = formatApplicationReference(appRef);
+
+  const journal: ExaminerWorkSchedule | null = await findJournal(staffNumber, null);
+
+  if (!journal) {
+    return createResponse(200);
```
Comment (TomBillingtonUK) [inline src/functions/searchBooking/framework/handler.ts]: Needs to be 404

File: src/functions/searchBooking/framework/handler.ts [hunk 4]
```diff
@@ -1,6 +1,98 @@
...
+  }
+
+  if (!journal.testSlots || journal.testSlots.length === 0) {
+    return createResponse(200);
```
Comment (TomBillingtonUK) [inline src/functions/searchBooking/framework/handler.ts]: Needs to be 404

File: src/functions/searchBooking/framework/handler.ts [hunk 5]
```diff
@@ -1,6 +1,98 @@
...
+  }
+
+  const testSlots = journal.testSlots
+    .map((testSlot) => {
+      if (testSlot.booking &&
+        testSlot.booking.application &&
+        testSlot.booking.application.applicationId &&
+        testSlot.booking.application.checkDigit &&
+        testSlot.booking.application.bookingSequence
```
Comment (TomBillingtonUK) [inline src/functions/searchBooking/framework/handler.ts]: lodash get function will do the same.

`get(testSlot, 'booking.application.specialNeeds', null);`

File: src/functions/searchBooking/framework/handler.ts [hunk 6]
```diff
@@ -1,6 +1,98 @@
...
+        if (parameterAppRef === formattedSlotAppRef) {
+          return testSlot;
+        }
+
+      }
+    })
+    .filter(testSlot => testSlot);
+
+  if (testSlots.length === 0) {
+    return createResponse(200);
```
Comment (TomBillingtonUK) [inline src/functions/searchBooking/framework/handler.ts]: should be a 404

**Review by TomBillingtonUK — APPROVED** [review]

File: package.json [hunk 7]
```diff
@@ -23,15 +23,18 @@
...
   },
   "dependencies": {
-    "@types/jwt-decode": "^2.2.1",
     "aws-sdk": "^2.389.0",
-    "moment": "^2.23.0"
+    "moment": "^2.23.0",
+    "@hapi/joi": "^15.0.3"
   },
   "devDependencies": {
     "@dvsa/mes-journal-schema": "^1.0.7",
```
Comment (rossfauldsbjss) [inline package.json]: Does journal schema need to be 1.1.0 for develop?

**Review by rossfauldsbjss — APPROVED** [review]
I think the journal schema needs to be 1.1.0 for develop now?
Apart from that looks good.

File: package.json (hunk 7 above)
Comment (amobarak) [reply package.json]: Good catch !

**Review by MarkPatt89 — APPROVED** [review]

**Pull request merged**